federation/pkg/federation-controller/service/dns/dns.go

```diff
@@ -0,0 +1,431 @@
+	for {
+		item, quit := d.workQueue.Get()
+		if quit {
+			return
+		}
+		service := item.(*v1.Service)
+		d.ensureDnsRecords(service)
+		d.workQueue.Done(item)
+	}
+}
+
+/* getResolvedEndpoints performs DNS resolution on the provided slice of endpoints (which might be DNS names or IPv4 addresses)
+   and returns a list of IPv4 addresses.  If any of the endpoints are neither valid IPv4 addresses nor resolvable DNS names,
+   non-nil error is also returned (possibly along with a partially complete list of resolved endpoints.
+*/
+func getResolvedEndpoints(endpoints []string) ([]string, error) {
+	resolvedEndpoints := make([]string, 0, len(endpoints))
+	for _, endpoint := range endpoints {
+		if net.ParseIP(endpoint) == nil {
+			// It's not a valid IP address, so assume it's a DNS name, and try to resolve it,
+			// replacing its DNS name with its IP addresses in expandedEndpoints
+			ipAddrs, err := net.LookupHost(endpoint)
+			if err != nil {
+				return resolvedEndpoints, err
+			}
+			resolvedEndpoints = append(resolvedEndpoints, ipAddrs...)
+
+		} else {
+			resolvedEndpoints = append(resolvedEndpoints, endpoint)
+		}
+	}
+	return resolvedEndpoints, nil
+}
+
+// getUplevelCName returns one level higher name for a dnsName, dnsNamePrefix is the common prefix after which zone/region string exist before the serviceDNSSuffix in a FQDN
+// if dnsName is of zone level, then return region CNAME, for e.g. if dnsName = "servicename.ns.federationname.svc.z1.r1.example.com", then return "servicename.ns.federationname.svc.r1.example.com"
+// if dnsName is of region level, then return global CNAME, for e.g. if dnsName = "servicename.ns.federationname.svc.r1.example.com", then return "servicename.ns.federationname.svc.example.com"
+// if dnsName is of global level, then return "" empty string, for e.g. if dnsName = "servicename.ns.federationname.svc.example.com", then return ""
+func (d *ServiceDNSController) getUplevelCName(dnsName, dnsNamePrefix string) (uplevelCname string) {
+	// split the dnsName with . as separator to get slice of strings
+	dnsNameSlice := strings.Split(dnsName, ".")
+	// find the index of the string which needs to be removed, so that we get uplevel CNAME
+	indexToDelete := len(strings.Split(dnsNamePrefix, "."))
+	// remove the zone/region from dnsName
+	dnsNameSlice = append(dnsNameSlice[:indexToDelete], dnsNameSlice[indexToDelete+1:]...)
+
+	uplevelCname = strings.Join(dnsNameSlice, ".")
+
+	// uplevelCname should have serviceDNSSuffix, otherwise the input dnsName was a global level name, so return empty string
+	if !strings.HasSuffix(uplevelCname, d.serviceDnsSuffix) {
+		uplevelCname = ""
+	}
+	return uplevelCname
+}
+
+/* ensureDnsRrsets ensures (idempotently, and with minimum mutations) that all of the DNS resource record sets are consistent with endpoints.
+ */
+func (d *ServiceDNSController) ensureDnsRrsets(dnsZone dnsprovider.Zone, endpoints map[string][]string, dnsNamePrefix string) error {
```

Comment:
I don't fully understand why you've changed the original code here.  Previously, whenever the state of a single service in a cluster changed, only that particular cluster's service was 'ensured' by calling this function.  Now there is no such cluster-specific function remaining, and all of the services in all of the clusters are ensured every time any one of them changes.  This seems like a lot of extra overhead, given that there could be hundreds of clusters (and services).

Why the change?